src/mappings/Rewards.ts

```diff
@@ -1,55 +1,72 @@
 import {HistoryElement} from '../types';
 import {SubstrateEvent} from "@subql/types";
-import {callsFromBatch, eventId, isBatch, distinct, timestamp} from "./common";
+import {callsFromBatch, eventIdFromBlockAndIdx, isBatch, distinct, timestamp, eventId} from "./common";
 import {CallBase} from "@polkadot/types/types/calls";
 import {AnyTuple} from "@polkadot/types/types/codec";
 
 function isPayoutStakers(call: CallBase<AnyTuple>): boolean {
     return call.method == "payoutStakers"
 }
 
-export async function handleReward(event: SubstrateEvent): Promise<void> {
-    const {event: {data: [account, newReward]}} = event;
+export async function handleReward(rewardEvent: SubstrateEvent): Promise<void> {
+    const {event: {data: [account, newReward]}} = rewardEvent;
+    let blockNumber = rewardEvent.block.block.header.number.toString()
+    let blockTimestamp = timestamp(rewardEvent.block)
 
-    const element = new HistoryElement(eventId(event));
+    let element = await HistoryElement.get(eventId(rewardEvent))
 
-    element.address = account.toString()
-    element.timestamp = timestamp(event.block)
+    if (element != undefined) {
+        // already processed reward previously
+        return;
+    }
 
-    const cause = event.extrinsic
+    const cause = rewardEvent.extrinsic
     const causeCall = cause.extrinsic.method
 
-    let validator;
+    let payoutedValidators: string[]
 
     if (isPayoutStakers(causeCall)) {
         const [validatorAddress,] = cause.extrinsic.method.args
 
-        validator = validatorAddress.toString()
+        payoutedValidators = [validatorAddress.toString()]
     } else if (isBatch(causeCall)) {
-        const validatorsInBatch = callsFromBatch(causeCall)
+        payoutedValidators = callsFromBatch(causeCall)
             .filter(isPayoutStakers)
             .map(payoutCall => payoutCall.args[0].toString()) // validator address
-
-        const distinctValidatorsInBatch = distinct(validatorsInBatch)
-
-        // cannot say which reward_stakers triggered this specific event,
-        // so can only be sure if there is only one distinct validator throughout all reward_stakers in batch
-        if (distinctValidatorsInBatch.length == 1) {
-            validator = distinctValidatorsInBatch[0]
-        } else {
-            validator = null
-        }
-    } else {
-        validator = null
     }
 
-    element.reward = {
-        amount: newReward.toString(),
-        isReward: true,
-        validator: validator
-    }
+    const distinctValidators = new Set(payoutedValidators)
+    let rewardEventType = api.events.staking.Reward;
 
-    await element.save();
+    let currentValidator;
+
+    const savingPromises = rewardEvent.block.events
+        .map((eventRecord, eventIndex) => {
```

Comment:
I would rewrite this through reduce where current state could be (currentValidator, listOfPromises). Just don't like combination of functional approach with side effects.

Reply:
yeah me neither

thanks for the idea with reduce

Reply:
Great!